pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,371 @@
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+
+	matchingPresets, err := filterPresets(allPresets, vm)
+
+	if err != nil {
+		logger.Object(vm).Reason(err).Errorf("Error while matching presets to VirtualMachine")
+		errors = append(errors, err)
+	}
+
+	if len(matchingPresets) != 0 {
+		err = applyPresets(vm, matchingPresets)
+		if err != nil {
+			// A more specific error should have been logged during the applyPresets call.
+			// We don't know *which* preset in the list was problematic at this level.
+			logger.Object(vm).Errorf("Unable to apply presets to virtualmachine: %v", err)
+			errors = append(errors, err)
+		}
+	}
+
+	logger.Object(vm).Info("Marking VM as initialized and updating")
+	removeInitializer(vm)
+	_, err = c.clientset.VM(vm.Namespace).Update(vm)
+	if err != nil {
+		logger.Object(vm).Errorf("Could not update VM. VM will not appear in the cluster: %v", err)
+		errors = append(errors, err)
+	}
+
+	if len(errors) > 0 {
+		return utilerrors.NewAggregate(errors)
+	}
+	return nil
+}
+
+// FIXME: There is probably a way to set up the vmPresetInformer such that
+// items are already partitioned into namespaces (and can just be listed)
+func listPresets(vmPresetInformer cache.SharedIndexInformer, namespace string) ([]kubev1.VirtualMachinePreset, error) {
+	result := []kubev1.VirtualMachinePreset{}
+	for _, obj := range vmPresetInformer.GetStore().List() {
+		preset := kubev1.VirtualMachinePreset{}
+		obj.(*kubev1.VirtualMachinePreset).DeepCopyInto(&preset)
+		if preset.Namespace == namespace {
+			result = append(result, preset)
+		}
+	}
+	return result, nil
+}
+
+// filterPresets returns list of VirtualMachinePresets which match given VirtualMachine.
+func filterPresets(list []kubev1.VirtualMachinePreset, vm *kubev1.VirtualMachine) ([]kubev1.VirtualMachinePreset, error) {
+	var matchingPresets []kubev1.VirtualMachinePreset
+
+	logger := log.Log
+
+	for _, preset := range list {
+		selector, err := k8smetav1.LabelSelectorAsSelector(&preset.Spec.Selector)
+		if err != nil {
+			logger.Reason(err).Errorf("label selector conversion failed: %v for selector: %v", preset.Spec.Selector, err)
+			return nil, fmt.Errorf("label selector conversion failed: %v for selector: %v", preset.Spec.Selector, err)
+		}
+
+		// check if the pod labels match the selector
+		if !selector.Matches(labels.Set(vm.Labels)) {
+			continue
+		}
+		logger.Infof("VirtualMachinePreset %s matches labels of VM %s", preset.GetName(), vm.GetName())
+		matchingPresets = append(matchingPresets, preset)
+	}
+	return matchingPresets, nil
+}
+
+func checkPresetMergeConflicts(presetSpec *kubev1.DomainSpec, vmSpec *kubev1.DomainSpec) error {
+	errors := []error{}
```

Comment:
Was it discussed that these values would be considered overrides if the VM spec has a preconfigured value before the preset is applied?  Basically, if a conflict occurs, the VM spec wins?

@fabiand  Do you recall this discussion at all?

Reply:
As I understand it, if the VM and the Preset disagree about a setting, the VM's setting should be used.

Reply:
> As I understand it, if the VM and the Preset disagree about a setting, the VM's setting should be used.

Is this an all or nothing situation.  For example, if part of a Preset conflicts, but other values in the Preset do not conflict, then do we take the values that are mergable or reject the entire Preset?

Reply:
Yes, it is all-or-nothing. This follows suit with the approach taken by Kubernetes with PodPresets, so hopefully follows the principle of least surprise.

Allowing partially applied presets also leads to questions about the meaning of annotations.